slotsmanager.c

```diff
@@ -0,0 +1,427 @@
+static void
+slots_gc_defrag(int ptype)
+{
+
+    if(!SlotsManager->info[ptype].enabled) return;
+    /* measure the fragmentation */
+    int i;
+    int frag = 0;
+    for(i = 1;
+        i < SlotsManager->info[ptype].size;
+        i ++) {
+
+        if( BASESLOT_PI(i, ptype)->gc.ReverseLink <
+            BASESLOT_PI(i - 1, ptype)->gc.ReverseLink
+            ) {
+            frag++;
+        }
+    }
+    int defrag = MPIU_Any(
+            frag > SlotsManager->info[ptype].size * 0.1,
+            MPI_COMM_WORLD);
+
+    /* if any rank is too fragmented, add a sorting to defrag. */
+    if(defrag)  {
+        /* sort the used ones
+         * by their location in the P array */
+        qsort_openmp(SlotsManager->info[ptype].ptr,
+                     SlotsManager->info[ptype].size,
+                     SlotsManager->info[ptype].elsize, 
+                     slot_cmp_reverse_link);
+    }
+}
+
+/* update new pointers. */
+static void
+slots_gc_collect(int ptype)
+{
+    int i;
+    if(!SlotsManager->info[ptype].enabled) return;
+
+    /* Now update the link in BhP */
+#pragma omp parallel for
+    for(i = 0;
+        i < SlotsManager->info[ptype].size;
+        i ++) {
+
+        P[BASESLOT_PI(i, ptype)->gc.ReverseLink].PI = i;
+    }
+}
+
+static int
+slots_gc_slots()
+{
+    int ptype;
+
+    int64_t total0[6];
+    int64_t total1[6];
+
+    int disabled = 1;
+    for(ptype = 0; ptype < 6; ptype ++) {
+        sumup_large_ints(1, &SlotsManager->info[ptype].size, &total0[ptype]);
+        if(total0[ptype] != 0) disabled = 0;
+    }
+    /* disabled this if no slots are used */
+    if(disabled) {
+        return 0;
+    }
+
+    slots_gc_mark();
+
+    for(ptype = 0; ptype < 6; ptype++) {
+        slots_gc_sweep(ptype);
+        slots_gc_defrag(ptype);
+        slots_gc_collect(ptype);
+    }
+
+#ifdef DEBUG
+    slots_check_id_consistency();
+#endif
+
+    for(ptype = 0; ptype < 6; ptype ++) {
+        sumup_large_ints(1, &SlotsManager->info[ptype].size, &total1[ptype]);
+
+        if(total1[ptype] != total0[ptype])
+            message(0, "GC: Reducing number of slots for %d from %ld to %ld\n", ptype, total0[ptype], total1[ptype]);
+    }
+
+    /* slot gc never invalidates the tree */
+    return 0;
+}
+
+
+void
+slots_reserve(int atleast[6])
+{
+    int newMaxSlots[6];
+    int ptype;
+    int good = 1;
+
+    if(SlotsManager->Base == NULL)
+        SlotsManager->Base = (char*) mymalloc("SlotsBase", 0);
+
+    for(ptype = 0; ptype < 6; ptype ++) {
+        newMaxSlots[ptype] = SlotsManager->info[ptype].maxsize;
+        while(newMaxSlots[ptype] < atleast[ptype]) {
+            int add = 0.2 * newMaxSlots[ptype];
+            if (add < 128) add = 128;
+            newMaxSlots[ptype] += add;
+            good = 0;
+        }
+    }
+    /* no need to grow, already have enough */
+    if (good) {
+        return;
+    }
+    /* FIXME: do a global max; because all variables in All.* are synced between ranks. */
```

Comment:
I don't think we want to fix this fixme: stars, SPH and BH can be inhomogeneous between ranks (because stars are very clustered). I was toying with the idea of one day allowing SPH memory to shrink sometimes to allow for this

Reply:
It was a left over. My fix was to move MaxSlots out of All. I think I removed the code to handle shrinking. in the case of shrinking the memmove have to be done in a different order.

Reply:
Oh, I see. That's fine.